File: mobile/modules/bluetooth-sdk/android/src/main/java/com/mentra/bluetoothsdk/drivers/GlassesDriverSgcAdapter.kt
package com.mentra.bluetoothsdk.drivers

import com.mentra.bluetoothsdk.sgcs.SGCManager

/**
 * Bridges a public [GlassesDriver] onto the internal [SGCManager] so the rest of
 * the app (DeviceManager, etc.) is unchanged. Public commands delegate to the
 * driver (respecting nothing here — capability gating is the caller's job for
 * now); app-internal/no-op members are supplied here so a driver never has to
 * implement them. See docs/device-driver-contract.md §5.
 */
class GlassesDriverSgcAdapter(
    private val driver: GlassesDriver,
    private val host: DeviceHost,
) : SGCManager() {

    init {
        type = driver.deviceType
        hasMic = driver.capabilities.hasMic
        host.reset()
        driver.start(host)
    }

    // ---- audio ----
    override fun setMicEnabled(enabled: Boolean) = driver.setMicEnabled(enabled)
    override fun sortMicRanking(list: MutableList<String>): MutableList<String> = driver.sortMicRanking(list)
    override suspend fun setImuEnabled(enabled: Boolean) = driver.setImuEnabled(enabled)

    // ---- camera / media ----
    override fun requestPhoto(
        requestId: String, appId: String, size: String, webhookUrl: String?, authToken: String?,
        compress: String?, flash: Boolean, save: Boolean, sound: Boolean, exposureTimeNs: Long?, iso: Int?,
    ) = driver.requestPhoto(requestId, appId, size, webhookUrl, authToken, compress, flash, save, sound, exposureTimeNs, iso)
    override fun startStream(message: MutableMap<String, Any>) = driver.startStream(message)
    override fun stopStream() = driver.stopStream()
    override fun sendStreamKeepAlive(message: MutableMap<String, Any>) = driver.sendStreamKeepAlive(message)
    override fun startVideoRecording(requestId: String, save: Boolean, flash: Boolean, sound: Boolean) =
        driver.startVideoRecording(requestId, save, flash, sound)
    override fun stopVideoRecording(requestId: String) = driver.stopVideoRecording(requestId)

    // ---- button settings (app-internal config; no-op) ----
    override fun sendButtonPhotoSettings() {}
    override fun sendButtonVideoRecordingSettings() {}
    override fun sendButtonMaxRecordingTime() {}
    override fun sendButtonCameraLedSetting() {}
    override fun sendCameraFovSetting() {}

    // ---- display ----
    override fun setBrightness(level: Int, autoMode: Boolean) = driver.setBrightness(level, autoMode)
    override fun clearDisplay() = driver.clearDisplay()
    override fun sendTextWall(text: String) = driver.sendTextWall(text)
    override fun sendDoubleTextWall(top: String, bottom: String) = driver.sendDoubleTextWall(top, bottom)
    override fun displayBitmap(base64ImageData: String, x: Int?, y: Int?, width: Int?, height: Int?): Boolean =
        driver.displayBitmap(base64ImageData, x, y, width, height)
    override fun showDashboard() = driver.showDashboard()
    override fun setDashboardPosition(height: Int, depth: Int) = driver.setDashboardPosition(height, depth)

    // ---- device control ----
    override fun setHeadUpAngle(angle: Int) = driver.setHeadUpAngle(angle)
    override fun getBatteryStatus() = driver.getBatteryStatus()
    override fun setSilentMode(enabled: Boolean) = driver.setSilentMode(enabled)
    override fun exit() = driver.clearDisplay()
    override fun sendShutdown() = driver.sendShutdown()
    override fun sendReboot() = driver.sendReboot()
    override fun sendRgbLedControl(
        requestId: String, packageName: String?, action: String, color: String?,
        onDurationMs: Int, offDurationMs: Int, count: Int,
    ) = driver.sendRgbLedControl(requestId, packageName, action, color, onDurationMs, offDurationMs, count)

    // ---- connection management ----
    override fun disconnect() = driver.disconnect()
    override fun forget() = driver.disconnect()
    override fun findCompatibleDevices() = driver.findCompatibleDevices()
    override fun stopScan() = driver.stopScan()
    override fun connectById(id: String) = driver.connectById(id)
    override fun getConnectedBluetoothName(): String = driver.getConnectedName() ?: ""
    override fun cleanup() = driver.cleanup()
    override fun ping() = driver.ping()
    override fun dbg1() {}
    override fun dbg2() {}

    // ---- network ----
    override fun requestWifiScan() = driver.requestWifiScan()
    override fun sendWifiCredentials(ssid: String, password: String) = driver.sendWifiCredentials(ssid, password)
    override fun forgetWifiNetwork(ssid: String) = driver.forgetWifiNetwork(ssid)
    override fun sendHotspotState(enabled: Boolean) = driver.sendHotspotState(enabled)

    // ---- user context / incident / gallery (app-internal; no-op) ----
    override fun sendUserEmailToGlasses(email: String) {}
    override fun sendIncidentId(incidentId: String, apiBaseUrl: String?) {}
    override fun queryGalleryStatus() {}
    override fun sendGalleryMode() {}

    // ---- version ----
    override fun requestVersionInfo() = driver.requestVersionInfo()
}

File: mobile/modules/bluetooth-sdk/ios/Source/drivers/GlassesDriverSgcAdapter.swift
//
//  GlassesDriverSgcAdapter.swift
//  Bridges a public GlassesDriver onto the internal SGCManager so the rest of
//  the app (DeviceManager, etc.) is unchanged. Public commands delegate to the
//  driver; app-internal/no-op members are supplied here. Mirror of the Android
//  GlassesDriverSgcAdapter.kt. See docs/device-driver-contract.md §5.
//

import Foundation

@MainActor
class GlassesDriverSgcAdapter: SGCManager {
    private let driver: GlassesDriver
    private let host: DeviceHost

    var type: String
    var hasMic: Bool

    init(driver: GlassesDriver, host: DeviceHost) {
        self.driver = driver
        self.host = host
        self.type = driver.deviceType
        self.hasMic = driver.capabilities.hasMic
        host.reset()
        driver.start(host)
    }

    // MARK: - Audio
    func setMicEnabled(_ enabled: Bool) { driver.setMicEnabled(enabled) }
    func sortMicRanking(list: [String]) -> [String] { driver.sortMicRanking(list: list) }
    func setImuEnabled(_ enabled: Bool) async { await driver.setImuEnabled(enabled) }

    // MARK: - Messaging
    func sendJson(_: [String: Any], wakeUp _: Bool, requireAck _: Bool) {}

    // MARK: - Camera & media
    func requestPhoto(_ requestId: String, appId: String, size: String?, webhookUrl: String?, authToken: String?, compress: String?, flash: Bool, save: Bool, sound: Bool, exposureTimeNs: Double?, iso: Int?) {
        driver.requestPhoto(requestId, appId: appId, size: size, webhookUrl: webhookUrl, authToken: authToken, compress: compress, flash: flash, save: save, sound: sound, exposureTimeNs: exposureTimeNs, iso: iso)
    }
    func startStream(_ message: [String: Any]) { driver.startStream(message) }
    func stopStream() { driver.stopStream() }
    func sendStreamKeepAlive(_ message: [String: Any]) { driver.sendStreamKeepAlive(message) }
    func startVideoRecording(requestId: String, save: Bool, flash: Bool, sound: Bool) {
        driver.startVideoRecording(requestId: requestId, save: save, flash: flash, sound: sound)
    }
    func stopVideoRecording(requestId: String) { driver.stopVideoRecording(requestId: requestId) }

    // MARK: - Button settings (app-internal config; no-op)
    func sendButtonPhotoSettings() {}
    func sendButtonVideoRecordingSettings() {}
    func sendButtonMaxRecordingTime() {}
    func sendButtonCameraLedSetting() {}
    func sendCameraFovSetting() {}

    // MARK: - Display
    func setBrightness(_ level: Int, autoMode: Bool) { driver.setBrightness(level, autoMode: autoMode) }
    func clearDisplay() { driver.clearDisplay() }
    func sendTextWall(_ text: String) async { await driver.sendTextWall(text) }
    func sendDoubleTextWall(_ top: String, _ bottom: String) async { await driver.sendDoubleTextWall(top, bottom) }
    func displayBitmap(base64ImageData: String, x: Int32?, y: Int32?, width: Int32?, height: Int32?) async -> Bool {
        await driver.displayBitmap(base64ImageData: base64ImageData, x: x, y: y, width: width, height: height)
    }
    func showDashboard() { driver.showDashboard() }
    func setDashboardPosition(_ height: Int, _ depth: Int) { driver.setDashboardPosition(height, depth) }

    // MARK: - Device control
    func setHeadUpAngle(_ angle: Int) { driver.setHeadUpAngle(angle) }
    func getBatteryStatus() { driver.getBatteryStatus() }
    func setSilentMode(_ enabled: Bool) { driver.setSilentMode(enabled) }
    func exit() { driver.clearDisplay() }
    func sendShutdown() { driver.sendShutdown() }
    func sendReboot() { driver.sendReboot() }
    func sendRgbLedControl(requestId: String, packageName: String?, action: String, color: String?, onDurationMs: Int, offDurationMs: Int, count: Int) {
        driver.sendRgbLedControl(requestId: requestId, packageName: packageName, action: action, color: color, onDurationMs: onDurationMs, offDurationMs: offDurationMs, count: count)
    }

    // MARK: - Connection management
    func disconnect() { driver.disconnect() }
    func forget() { driver.disconnect() }
    func findCompatibleDevices() { driver.findCompatibleDevices() }
    func stopScan() { driver.stopScan() }
    func connectById(_ id: String) { driver.connectById(id) }
    func connectController() {}
    func disconnectController() {}
    func getConnectedBluetoothName() -> String? { driver.getConnectedName() }
    func cleanup() { driver.cleanup() }
    func ping() { driver.ping() }
    func dbg1() {}
    func dbg2() {}

    // MARK: - Network management
    func requestWifiScan() { driver.requestWifiScan() }
    func sendWifiCredentials(_ ssid: String, _ password: String) { driver.sendWifiCredentials(ssid, password) }
    func forgetWifiNetwork(_ ssid: String) { driver.forgetWifiNetwork(ssid) }
    func sendHotspotState(_ enabled: Bool) { driver.sendHotspotState(enabled) }
    func sendOtaStart() {}
    func sendOtaQueryStatus() {}
    func sendUserEmailToGlasses(_: String) {}
    func sendIncidentId(_: String, apiBaseUrl _: String?) {}

    // MARK: - Gallery / version
    func queryGalleryStatus() {}
    func sendGalleryMode() {}
    func requestVersionInfo() { driver.requestVersionInfo() }
}
